fairscale/experimental/nn/distributed_pipeline/graph.py

```diff
@@ -0,0 +1,202 @@
+        """
+        self._set_inputs(module, [DataSource(None, ind)])
+
+    def add_multi_input_layer(self, module: RemoteModule, inputs: List[RemoteModule]) -> None:
+        """Adds a module with multiple inputs to the graph. The modules that provide inputs to this module
+        must have been added previously to the graph and are listed with argument inputs.
+        """
+        self._set_inputs(module, [DataSource(self._find_node(m), 0) for m in inputs])
+
+    def fan_out(self, module: RemoteModule, outputs: List[RemoteModule]) -> None:
+        """Feeds outputs of a previously added module to modules specified by argument 'outputs' (so
+        'module' should have at least 'len(outputs)' outputs.
+        Modules in the list 'outputs' are added to the graph if they have not been added previously.
+        """
+        node = self._find_node(module)
+        node.num_outputs = len(outputs)
+        for i, m in enumerate(outputs):
+            self._set_inputs(m, [DataSource(node, i)])
+
+    def _compile(self) -> None:
+        """Precomputes self.model_input_consumers and self.output_consumers for internal use by the pipleine
+        class. These two lists show consumers of inputs to the model, and outputs of each module of
+        the graph. Each consumer is a pair (i, j) which stands for the j'th input to the i'th module
+        in the graph.
+        """
+
+        # TODO: We need to make sure following conditions hold before preparing the graph for the pipeline:
+        #   * the graph has a least one module, and is connected.
+        #   * num_inputs and num_outputs for modules matche list of connections defined in the graph.
+        #   * all inputs to a module should come from model input, or modules with smaller index in
+        #     the graph. This condition is used in implementaion of DistributedPipeline.forward. Even
+        #     if we relax this condition, still need to make sure the graph is acyclic.
+
+        m = len(self.nodes)
+        self.model_input_consumers = []
+        for node in self.nodes:
+            for input_index, input_item in enumerate(node.inputs):
+                data_consumer = NodeDataConsumer(node, input_index, input_item.output_idx)
+                if input_item.producer is not None:
+                    input_item.producer.output_consumers.append(data_consumer)
+                else:
+                    self.model_input_consumers.append(data_consumer)
+
+    def _trace_modules(self, node: Node) -> List[Node]:
+        """Compiles a list of modules (starting from module number module_idx), where each module in the list
+        gets the output of previous module in the list as its input. So every module in the list, except the
+        first one should have only one input, and similarly, every module in the list, except the last one
+        should have only one output.
+        """
+        partition = []
```

Comment:
That's because the way each partition is going to work. It waits untill all required inputs are ready and then it runs the partition. This may endup in a deadlock if partitions can have arbitrary connections to each other. 
For example, consider modules m1, m2, m3, where m1 and m3 are on device0 and m2 is  on device1, and iinput connections are as fllow:
m1 --> m2
m1 --> m3
m2 --> m3
 If we create have m1 and m3 on the same partition and m2 on another partition, there will be a deadlock.